File: 2019/knowit/19/19.hs
-- Note: Digits are in reverse order
type Digits = [Int]

toDigits :: Int -> Digits
toDigits 0 = []
toDigits n = n `mod` 10 : toDigits (n `div` 10)

fromDigits :: Digits -> Int
fromDigits ds = fromDigits' 0 ds
    where fromDigits' _ [] = 0
          fromDigits' n (d:ds) = d*10^n + fromDigits' (n+1) ds

isHiddenPalindrome :: Int -> Bool
isHiddenPalindrome n = isPalindrome (n + (fromDigits . reverse . toDigits $ n))

isPalindrome :: Int -> Bool
isPalindrome n = ds == reverse ds
    where ds = toDigits n

main =
    print
        . sum
        . filter (not . isPalindrome)
        . filter isHiddenPalindrome
        $ [1..123454321]

File: 2019/knowit/19/19.js
"use strict";

function* nums(start, end) {
	let i = start;
	do {
		yield i++;
	} while (i <= end);
}

const isPalindrome = n => {
	const s = n.toString();
	const l = s.length;
	const lh = l / 2;
	for (let i = 0; i < lh; i++) {
		if (s.charAt(i) !== s.charAt(l-1-i)) {
			return false;
		}
	}
	return true;
};

const isHiddenPalindrome = n => isPalindrome(Number.parseInt(n.toString().split("").reverse().join("")) + n);

let sum = 0;
for (const n of nums(1, 123454321)) {
	if (isHiddenPalindrome(n) && !isPalindrome(n)) {
		sum += n;
	}
}

console.log(sum);
